app/slave/cluster_slave.py

```diff
@@ -74,24 +75,36 @@ def _start_heartbeat(self):
 
     def _run_heartbeat(self):
         try:
-            self._send_heartbeat_to_master()
-            self._heartbeat_failure_count = 0
+            response = self._send_heartbeat_to_master()
+
+            # A non HTTP 200 response indicates that the master is up, but does not recognize the slave.
+            # The slave tries to reconnect to master at this point.
+            # This is also counted as a heartbeat failure and the slave takes an action accordingly
+            if response.status_code != http.client.OK:
+                self._act_on_heartbeat_failure()
+                self._logger.info('Master responded with {}. Attempting to reconnect.'.format(response.status_code))
+                self.connect_to_master(self._master_url)
+            else:
+                self._heartbeat_failure_count = 0
         except (requests.ConnectionError, requests.Timeout):
-            self._heartbeat_failure_count += 1
-            if self._heartbeat_failure_count >= self._heartbeat_failure_threshold:
-                self._logger.error('Master is not responding to heartbeats')
-
-                # TODO: Right now the slave simply dies when it does not hear back from master. The next step would
-                # be to try to reconnect to master at this point. In future the heartbeat and connect_to_master
-                # methods can combined into one. This combined method will behave differently based on current state.
-                self.kill()
+            self._act_on_heartbeat_failure()
 
         self._hb_scheduler.enter(self._heartbeat_interval, 0, self._run_heartbeat)
 
+    def _act_on_heartbeat_failure(self):
+        self._heartbeat_failure_count += 1
+        if self._heartbeat_failure_count >= self._heartbeat_failure_threshold:
+            self._logger.error('Master is not responding to heartbeats')
+
+            # TODO: Right now the slave simply dies when it does not hear back from master. The next step would
+            # be to try to reconnect to master at this point. In future the heartbeat and connect_to_master
+            # methods can combined into one. This combined method will behave differently based on current state.
+            self.kill()
```

Comment:
If this change is connecting slaves to the master, then I think it's time to get rid of this `self.kill()`

as per this comment
https://github.com/box/ClusterRunner/pull/422#discussion_r180558189

Reply:
Which further implies that it's time to get rid of this method completely